### PR TITLE
Further fixes for performance with internal bucketing

### PR DESCRIPTION
**Pull request by puneeshkhanna**
* Further fixes for performance with internal bucketing.

Also add clear cache() to save memory.
make style changes also added.

* Calculate kv cache sliding idx for the decode phase only.

* Add hpu graphs check for clear cache.

---------

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by puneeshkhanna** [comment]
@regisss -
Please first merge https://github.com/huggingface/optimum-habana/pull/720 followed by this PR.
This PR has additional enhancements over https://github.com/huggingface/optimum-habana/pull/720.

Assume input tokens as 128 and new tokens as 512. 
Without bucketing and slicing changes in these 2 PRs, in the decode phase we used to calculate the attention scores by multiplying with full KV cache of size (512+128).

Now with the changes of these 2 PRs, we l consider KV caches as below:
1. Decode phases of tokens 128-256 -> sliced KV cache till seq len 256.
2. Decode phases of tokens 256-384 -> sliced KV cache till seq len 384.
3. Decode phases of tokens 384-512 -> sliced KV cache till seq len 512.
4. And so on.

The bucketing changes along with reuse cache gives enhanced performance. See improved performances in below table.
image

For llama runs-> --bucket_internal --bucket_size 128 will be the default suggested arguments in README for all configs.

**Comment by puneeshkhanna** [comment]
@x574chen  - please check the enhanced changes in this PR over your main changes.

**Review by dvarshney-habana — APPROVED** [review]

**Comment by puneeshkhanna** [comment]
@regisss  - please trigger CI.

File: optimum/habana/transformers/generation/utils.py [hunk 1]
```diff
@@ -678,6 +681,7 @@ def generate(
         model_kwargs["attn_softmax_bf16"] = generation_config.attn_softmax_bf16
 
         # determine whether limit_hpu_graphs needs to be used
+        model_kwargs["use_hpu_graphs"] = hpu_graphs
```
Comment (xt574chen) [inline optimum/habana/transformers/generation/utils.py:684]: Is it still needed after clear_cache removed?

**Review by xt574chen — APPROVED** [review]
The changes to token_idx_cpu look good to me.

File: optimum/habana/transformers/generation/utils.py (hunk 1 above)
Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/utils.py:684]: Lets have it ; no harm as we may need to use it in future.

**Comment by puneeshkhanna** [comment]
@regisss  - Can we merge this ?

**Review by regisss — APPROVED** [review]
LGTM!

The code style check failed, it's probably due to the new version of `ruff`. Could you rebase your branch on main and then run the following please?
```
pip install -U ruff
make style
```

**Comment by puneeshkhanna** [comment]
@regisss  - I am not seeing any "make style" issues with the files that I changed. 
We should fix that in separate PR for the files that require style updates.

**Comment by puneeshkhanna** [comment]
image
@regisss  - FYI.

**Comment by regisss** [comment]
> @regisss - I am not seeing any "make style" issues with the files that I changed. We should fix that in separate PR for the files that require style updates.

It's due to a new version of `ruff` and has been updated in the main branch. Please rebase on main and run `make style` again, that should make the CI pass.

**Comment by puneeshkhanna** [comment]
@regisss  - please review and merge https://github.com/huggingface/optimum-habana/pull/781 instead of this. I m discarding this PR.

**Pull request closed without merging**